src/traefik.py

```diff
@@ -0,0 +1,616 @@
+        service_def = {
+            "loadBalancer": lb_def,
+        }
+
+        if is_reverse_termination:
+            # i.e. traefik itself is not related to tls certificates, but the ingress requirer is
+            transport_name = "reverseTerminationTransport"
+            lb_def["serversTransport"] = transport_name
+            transports = {transport_name: {"insecureSkipVerify": False}}
+
+        elif is_termination:
+            # i.e. traefik itself is related to tls certificates, but the ingress requirer is not
+            transports = {}
+
+        elif is_end_to_end:
+            # We cannot assume traefik's CA is the same CA that signed the proxied apps.
+            # Since we use the update_ca_certificates machinery, we don't need to specify the
+            # "rootCAs" entry.
+            # Keeping the serverTransports section anyway because it is informative ("endToEndTLS"
+            # vs "reverseTerminationTransport") when inspecting the config file in production.
+            transport_name = "endToEndTLS"
+            lb_def["serversTransport"] = transport_name
+            transports = {transport_name: {"insecureSkipVerify": False}}
+
+        else:
+            transports = {}
+
+        config = {
+            "http": {
+                "routers": router_cfg,
+                "services": {traefik_service_name: service_def},
+            },
+        }
+        # Traefik does not accept an empty serversTransports. Add it only if it's non-empty.
+        if transports:
+            config["http"].update({"serversTransports": transports})
+
+        middlewares = self._generate_middleware_config(
+            redirect_https=redirect_https_,
+            strip_prefix=strip_prefix_,
+            scheme=scheme_,
+            prefix=prefix,
+        )
+
+        if middlewares:
+            config["http"]["middlewares"] = middlewares
+            router_cfg[traefik_router_name]["middlewares"] = list(middlewares.keys())
+
+            if f"{traefik_router_name}-tls" in router_cfg:
+                router_cfg[f"{traefik_router_name}-tls"]["middlewares"] = list(middlewares.keys())
+
+        return config
+
+    def _generate_middleware_config(
+        self,
+        redirect_https: bool,
+        strip_prefix: bool,
+        scheme: str,
+        prefix: str,
+    ) -> dict:
+        """Generate a middleware config.
+
+        We need to generate a different section per middleware type, otherwise traefik complains:
+          "cannot create middleware: multi-types middleware not supported, consider declaring two
+          different pieces of middleware instead"
+        """
+        no_prefix_middleware = {}  # type: Dict[str, Dict[str, Any]]
+        if self._routing_mode is RoutingMode.path:
+            if strip_prefix:
+                no_prefix_middleware[f"juju-sidecar-noprefix-{prefix}"] = {
+                    "stripPrefix": {"prefixes": [f"/{prefix}"], "forceSlash": False}
+                }
```

Comment:
I would remove one indentation level:

```suggestion
        if self._routing_mode is RoutingMode.path and strip_prefix:
            no_prefix_middleware[f"juju-sidecar-noprefix-{prefix}"] = {
                "stripPrefix": {"prefixes": [f"/{prefix}"], "forceSlash": False}
            }
```

Reply:
good thinking